File: Anytime_Planning/brentsc-infoplanner/src/robot.cpp
//
// Created by brent on 4/7/17.
//

#include <igl/robot.h>
#include <igl/utils/utils_nx.h>
#include <iostream>
#include <set>
#include <igl/se3_pose.h>


template<class state>
void nx::Robot<state>::applyControl(std::vector<int> &action_idx, const int n_controls) {
  // If the robot has no actions, give a warning.
  if (action_idx.size() == 0) {
    std::cout << "WARNING: No control inputs available for robot to apply.\n";
  }

  std::vector<state> next_micro;
  int i = 0;
  for (auto it = action_idx.rbegin(); it != action_idx.rend() && i < n_controls; ++it, i++) {
    this->env->forward_action(this->getState(), action_idx.back(), next_micro);
    this->x = next_micro.back(); // Move the robot's position.
    action_idx.pop_back();
  }
}

template<class state>
state nx::Robot<state>::getState() const {
  return x;
}

template<class state>
void nx::Robot<state>::setState(state s) {
  x = s;
}

// Explicit Instantiations

template
class nx::Robot<std::vector<int>>;

template
class nx::Robot<nx::SE3Pose>;

File: Anytime_Planning/brentsc-infoplanner/src/robot.py
from typing import List
import igl.robot
import igl.utils.utils_nx
import igl.se3_pose


class Robot:
    def __init__(self, env, x):
        self.env = env
        self.x = x

    def applyControl(self, action_idx: List[int], n_controls: int):
        if not action_idx:
            print("WARNING: No control inputs available for robot to apply.")
            return

        next_micro = []
        i = 0
        for it in reversed(action_idx):
            if i >= n_controls:
                break
            self.env.forward_action(self.getState(), action_idx[-1], next_micro)
            self.x = next_micro[-1]  # Move the robot's position.
            action_idx.pop()

    def getState(self):
        return self.x

    def setState(self, s):
        self.x = s
